tests/tape/interfaces/test_qnode_jax.py

```diff
@@ -0,0 +1,205 @@
+    """Transforms every RY(y) gate in a circuit to RX(-a*cos(y))"""
+
+    def construct(self, args, kwargs):
+        """New quantum tape construct method, that performs
+        the transform on the tape in a define-by-run manner"""
+
+        # the following global variable is defined simply for testing
+        # purposes, so that we can easily extract the transformed operations
+        # for verification.
+
+        t_op = []
+
+        QNode.construct(self, args, kwargs)
+
+        new_ops = []
+        for o in self.qtape.operations:
+            # here, we loop through all tape operations, and make
+            # the transformation if a RY gate is encountered.
+            if isinstance(o, qml.RY):
+                t_op.append(qml.RX(-a * framework.cos(o.data[0]), wires=o.wires))
+                new_ops.append(t_op[-1])
+            else:
+                new_ops.append(o)
+
+        self.qtape._ops = new_ops
+        self.qtape._update()
+
+    import copy
+
+    new_qnode = copy.deepcopy(qnode)
+    new_qnode.construct = construct.__get__(new_qnode, QNode)
+    return new_qnode
+
+
+@pytest.mark.parametrize(
+    "dev_name,diff_method",
+    [("default.mixed", "finite-diff"), ("default.qubit.autograd", "parameter-shift")],
+)
+def test_transform(dev_name, diff_method, monkeypatch, tol):
+    """Test an example transform"""
+    monkeypatch.setattr(qml.operation.Operation, "do_check_domain", False)
+
+    dev = qml.device(dev_name, wires=1)
+
+    @qnode(dev, interface="jax", diff_method=diff_method)
+    def circuit(weights):
+        # the following global variables are defined simply for testing
+        # purposes, so that we can easily extract the operations for verification.
+        op1 = qml.RY(weights[0], wires=0)
+        op2 = qml.RX(weights[1], wires=0)
+        return qml.expval(qml.PauliZ(wires=0))
+
+    weights = np.array([0.32, 0.543])
+    a = np.array(0.5)
+
+    def loss(weights, a):
+        # the following global variable is defined simply for testing
+        # purposes, so that we can easily extract the transformed QNode
+        # for verification.
+
+        # transform the circuit QNode with trainable weight 'a'
+        new_circuit = qtransform(circuit, a)
+
+        # evaluate the transformed QNode
+        res = new_circuit(weights)
+
+        # evaluate the original QNode with pre-processed parameters
+        res2 = circuit(jnp.sin(weights))
+
+        # return the sum of the two QNode evaluations
+        return res + res2
+
+    res = loss(weights, a)
+
+    grad = jax.grad(loss, argnums=[0, 1])(weights, a)
+    assert len(grad) == 2
+    assert grad[0].shape == weights.shape
+    assert grad[1].shape == a.shape
+
+    # compare against the expected values
+    tol = 1e-5
+    assert np.allclose(res, 1.8244501889992706, atol=tol, rtol=0)
+    assert np.allclose(grad[0], [-0.26610258, -0.47053553], atol=tol, rtol=0)
+    assert np.allclose(grad[1], 0.06486032, atol=tol, rtol=0)

```

Comment:
Would it be more unittest-like to just apply the transform and check the queue of the new qnode, rather than using a very elaborate post-processing function and checking some pre-computed numerical values?

Reply:
I would disagree. The purpose of this check is to ensure that the `jax.grad` transform executes properly with the tape transform. The thing I specifically want to only check here is the numerics. If we only check the queue, it will miss the code path I'm aiming at.